File: datasets/C++/Qwen/Algorithm/hard2.cpp
#include <iostream>
#include <vector>
#include <cmath>
using namespace std;

int chiefHopper(vector<int>& arr) {
    double required = 0;
    for (int i = arr.size() - 1; i >= 0; i--) {
        required = ceil((required + arr[i]) / 2.0);
    }
    return (int)required;
}

int main() {
    int n;
    cin >> n;
    vector<int> arr(n);
    for (int i = 0; i < n; i++) {
        cin >> arr[i];
    }
    cout << chiefHopper(arr) << endl;
    return 0;
}

File: datasets/Python/Qwen/Algorithm/hard2.py
import math

def chiefHopper(arr):
    required = 0
    for height in reversed(arr):
        required = math.ceil((required + height) / 2)
    return required

# Manual Input
n = int(input())
arr = list(map(int, input().split()))

print(chiefHopper(arr))